packages/xo-server/src/xo-mixins/pool.js

```diff
@@ -0,0 +1,77 @@
+import { difference, uniq } from 'lodash'
+
+export default class {
+  constructor(xo) {
+    this._xo = xo
+  }
+
+  async mergeInto({ sources, target, force }) {
+    const { _xo } = this
+    const targetHost = _xo.getObject(target.master)
+    const _sources = []
+    const sourceHosts = []
+    let source
+    let sourceHost
+    for (const sourceId of sources) {
+      source = _xo.getObject(sourceId)
+      sourceHost = _xo.getObject(source.master)
+      if (sourceHost.productBrand !== targetHost.productBrand) {
+        throw new Error(
+          `a ${sourceHost.productBrand} pool cannot be merged into a ${
+            targetHost.productBrand
+          } pool`
+        )
+      }
+      _sources.push(source)
+      sourceHosts.push(sourceHost)
+    }
+
+    const targetRequiredPatches = uniq(
+      Array.prototype.concat.call(
+        ...(await Promise.all(
+          _sources.map(({ master }) =>
+            _xo.getPatchesDifference(master, target.master)
+          )
+        ))
+      )
+    )
+
+    const promises = []
+    if (targetRequiredPatches.length > 0) {
+      const targetXapi = _xo.getXapi(target)
+      promises.push(
+        targetXapi.installPatches({
+          patches: await targetXapi.findPatches(targetRequiredPatches),
+        })
+      )
+    }
+
+    const allRequiredPatches = targetRequiredPatches.concat(
+      targetHost.patches.map(patchId => _xo.getObject(patchId).name)
+    )
+
+    let sourcePatches
+    let sourceRequiredPatches
+    let sourceXapi
+    for (sourceHost of sourceHosts) {
+      sourcePatches = sourceHost.patches.map(
+        patchId => _xo.getObject(patchId).name
+      )
+      sourceRequiredPatches = difference(allRequiredPatches, sourcePatches)
+      if (sourceRequiredPatches.length > 0) {
+        sourceXapi = _xo.getXapi(sourceHost.$pool)
+        promises.push(
+          sourceXapi.installPatches({
+            patches: await sourceXapi.findPatches(sourceRequiredPatches),
+          })
+        )
+      }
+    }
+
+    await Promise.all(promises)
+
+    for (source of _sources) {
```

Comment:
You can use `sourceHosts` so you don't need a `_sources` collection.

Reply:
I need `source._xapiId`.